File: engine/src/agent33/api/routes/explanations.py
"""FastAPI router for explanation generation and management."""

from __future__ import annotations

import uuid
from typing import Any

import structlog
from fastapi import APIRouter, HTTPException, Request

from agent33.explanation.fact_check import run_fact_check_hooks
from agent33.explanation.models import (
    DiffReviewRequest,
    ExplanationClaim,
    ExplanationMetadata,
    ExplanationMode,
    ExplanationRequest,
    PlanReviewRequest,
    ProjectRecapRequest,
)
from agent33.explanation.renderer import (
    render_diff_review,
    render_plan_review,
    render_project_recap,
)
from agent33.explanation.store import ExplanationStore
from agent33.security.permissions import require_scope

logger = structlog.get_logger()

router = APIRouter(prefix="/v1/explanations", tags=["explanations"])

# Module-level fallback store used when no app.state store is configured.
# Routes prefer app.state.explanation_store when available so that the store
# path can be configured at startup (and swapped to :memory: in tests).
_default_store: ExplanationStore | None = None


def _get_store(request: Request | None = None) -> ExplanationStore:
    """Return the active ExplanationStore.

    Precedence:
    1. ``request.app.state.explanation_store`` (set during lifespan or test fixture)
    2. The module-level ``_default_store`` singleton (lazy-initialised on first call)
    """
    global _default_store  # noqa: PLW0603

    if request is not None:
        store: ExplanationStore | None = getattr(request.app.state, "explanation_store", None)
        if store is not None:
            return store

    if _default_store is None:
        _default_store = ExplanationStore()
    return _default_store


def _build_explanation(
    explanation_id: str,
    entity_type: str,
    entity_id: str,
    mode: ExplanationMode,
    content: str,
    metadata: dict[str, Any],
    claim_requests: list[Any],
) -> ExplanationMetadata:
    return ExplanationMetadata(
        id=explanation_id,
        entity_type=entity_type,
        entity_id=entity_id,
        mode=mode,
        content=content,
        metadata=metadata,
        claims=[
            ExplanationClaim(
                claim_type=claim.claim_type,
                target=claim.target,
                expected=claim.expected,
                description=claim.description,
            )
            for claim in claim_requests
        ],
    )


@router.post("/", dependencies=[require_scope("workflows:write")], status_code=201)
async def create_explanation(
    request_body: ExplanationRequest, request: Request
) -> ExplanationMetadata:
    """Generate a new explanation scaffold for an entity."""
    explanation_id = f"expl-{uuid.uuid4().hex[:12]}"
    explanation = _build_explanation(
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        mode=request_body.mode,
        content=_render_explanation_content(request_body),
        metadata=request_body.metadata,
        claim_requests=request_body.claims,
    )

    explanation.fact_check_status = await run_fact_check_hooks(explanation)
    _get_store(request).save(explanation)

    logger.info(
        "explanation_created",
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        mode=request_body.mode,
        claims=len(explanation.claims),
        fact_check_status=explanation.fact_check_status,
    )
    return explanation


@router.get("/{explanation_id}", dependencies=[require_scope("workflows:read")])
async def get_explanation(explanation_id: str, request: Request) -> ExplanationMetadata:
    """Retrieve an explanation by ID."""
    explanation = _get_store(request).get(explanation_id)
    if explanation is None:
        raise HTTPException(status_code=404, detail=f"Explanation '{explanation_id}' not found")

    logger.info("explanation_retrieved", explanation_id=explanation_id)
    return explanation


@router.get("/", dependencies=[require_scope("workflows:read")])
async def list_explanations(
    request: Request,
    entity_type: str | None = None,
    entity_id: str | None = None,
) -> list[ExplanationMetadata]:
    """List explanations, optionally filtered by entity."""
    results = _get_store(request).list(entity_type=entity_type, entity_id=entity_id)

    logger.info(
        "explanations_listed",
        count=len(results),
        entity_type=entity_type,
        entity_id=entity_id,
    )
    return results


@router.post(
    "/{explanation_id}/fact-check",
    dependencies=[require_scope("workflows:write")],
)
async def rerun_fact_check(explanation_id: str, request: Request) -> ExplanationMetadata:
    """Re-run deterministic fact-check validation for an explanation."""
    store = _get_store(request)
    explanation = store.get(explanation_id)
    if explanation is None:
        raise HTTPException(status_code=404, detail=f"Explanation '{explanation_id}' not found")

    explanation.fact_check_status = await run_fact_check_hooks(explanation)
    store.save(explanation)

    logger.info(
        "explanation_fact_check_rerun",
        explanation_id=explanation_id,
        fact_check_status=explanation.fact_check_status,
    )
    return explanation


@router.get(
    "/{explanation_id}/claims",
    dependencies=[require_scope("workflows:read")],
)
async def get_explanation_claims(explanation_id: str, request: Request) -> list[ExplanationClaim]:
    """Retrieve deterministic fact-check claims for an explanation."""
    explanation = _get_store(request).get(explanation_id)
    if explanation is None:
        raise HTTPException(status_code=404, detail=f"Explanation '{explanation_id}' not found")
    return explanation.claims


@router.delete("/{explanation_id}", dependencies=[require_scope("workflows:write")])
async def delete_explanation(explanation_id: str, request: Request) -> dict[str, str]:
    """Delete an explanation."""
    deleted = _get_store(request).delete(explanation_id)
    if not deleted:
        raise HTTPException(status_code=404, detail=f"Explanation '{explanation_id}' not found")

    logger.info("explanation_deleted", explanation_id=explanation_id)
    return {"message": f"Explanation '{explanation_id}' deleted"}


def get_explanations_store(request: Request | None = None) -> ExplanationStore:
    """Return the active store (used by test fixtures)."""
    return _get_store(request)


@router.post(
    "/diff-review",
    dependencies=[require_scope("workflows:write")],
    status_code=201,
)
async def create_diff_review(
    request_body: DiffReviewRequest, request: Request
) -> ExplanationMetadata:
    """Generate a diff review visual page."""
    explanation_id = f"expl-{uuid.uuid4().hex[:12]}"

    html_content = render_diff_review(
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        diff_text=request_body.diff_text,
    )

    explanation = _build_explanation(
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        mode=ExplanationMode.DIFF_REVIEW,
        content=html_content,
        metadata=request_body.metadata,
        claim_requests=request_body.claims,
    )

    explanation.fact_check_status = await run_fact_check_hooks(explanation)
    _get_store(request).save(explanation)

    logger.info(
        "diff_review_created",
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        claims=len(explanation.claims),
        fact_check_status=explanation.fact_check_status,
    )
    return explanation


@router.post(
    "/plan-review",
    dependencies=[require_scope("workflows:write")],
    status_code=201,
)
async def create_plan_review(
    request_body: PlanReviewRequest, request: Request
) -> ExplanationMetadata:
    """Generate a plan review visual page."""
    explanation_id = f"expl-{uuid.uuid4().hex[:12]}"

    html_content = render_plan_review(
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        plan_text=request_body.plan_text,
    )

    explanation = _build_explanation(
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        mode=ExplanationMode.PLAN_REVIEW,
        content=html_content,
        metadata=request_body.metadata,
        claim_requests=request_body.claims,
    )

    explanation.fact_check_status = await run_fact_check_hooks(explanation)
    _get_store(request).save(explanation)

    logger.info(
        "plan_review_created",
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        claims=len(explanation.claims),
        fact_check_status=explanation.fact_check_status,
    )
    return explanation


@router.post(
    "/project-recap",
    dependencies=[require_scope("workflows:write")],
    status_code=201,
)
async def create_project_recap(
    request_body: ProjectRecapRequest, request: Request
) -> ExplanationMetadata:
    """Generate a project recap visual page."""
    explanation_id = f"expl-{uuid.uuid4().hex[:12]}"

    html_content = render_project_recap(
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        recap_text=request_body.recap_text,
        highlights=request_body.highlights,
    )

    explanation = _build_explanation(
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        mode=ExplanationMode.PROJECT_RECAP,
        content=html_content,
        metadata=request_body.metadata,
        claim_requests=request_body.claims,
    )

    explanation.fact_check_status = await run_fact_check_hooks(explanation)
    _get_store(request).save(explanation)

    logger.info(
        "project_recap_created",
        explanation_id=explanation_id,
        entity_type=request_body.entity_type,
        entity_id=request_body.entity_id,
        claims=len(explanation.claims),
        fact_check_status=explanation.fact_check_status,
    )
    return explanation


def _render_explanation_content(request: ExplanationRequest) -> str:
    title = f"{request.mode.value.replace('_', ' ').title()} explanation"
    sections = [
        title,
        f"Entity: {request.entity_type} '{request.entity_id}'",
    ]
    highlights = request.metadata.get("highlights", [])
    if isinstance(highlights, list) and highlights:
        sections.append("Highlights:")
        sections.extend(f"- {item}" for item in highlights if isinstance(item, str))
    return "\n".join(sections)

File: frontend/src/data/domains/explanations.ts
import type { DomainConfig } from "../../types";

export const explanationsDomain: DomainConfig = {
  id: "explanations",
  title: "Explanations",
  description: "Generate and manage explanations with fact-check validation.",
  operations: [
    {
      id: "explanations-create",
      title: "Create Explanation",
      method: "POST",
      path: "/v1/explanations/",
      description: "Generate a new explanation for an entity.",
      defaultBody: JSON.stringify(
        {
          entity_type: "workflow",
          entity_id: "hello-flow",
          mode: "plan_review",
          metadata: {
            model: "llama3.1",
            highlights: ["Review generated plan and constraints"]
          },
          claims: [
            {
              claim_type: "metadata_equals",
              target: "model",
              expected: "llama3.1",
              description: "Expected model metadata must be preserved"
            }
          ]
        },
        null,
        2
      )
    },
    {
      id: "explanations-get",
      title: "Get Explanation",
      method: "GET",
      path: "/v1/explanations/{explanation_id}",
      description: "Retrieve an explanation by ID.",
      defaultPathParams: {
        explanation_id: "expl-abc123"
      }
    },
    {
      id: "explanations-list",
      title: "List Explanations",
      method: "GET",
      path: "/v1/explanations/",
      description: "List all explanations with optional filters."
    },
    {
      id: "explanations-list-by-entity",
      title: "List by Entity",
      method: "GET",
      path: "/v1/explanations/",
      description: "List explanations filtered by entity.",
      defaultQuery: {
        entity_type: "workflow",
        entity_id: "hello-flow"
      }
    },
    {
      id: "explanations-delete",
      title: "Delete Explanation",
      method: "DELETE",
      path: "/v1/explanations/{explanation_id}",
      description: "Delete an explanation.",
      defaultPathParams: {
        explanation_id: "expl-abc123"
      }
    },
    {
      id: "explanations-rerun-fact-check",
      title: "Re-run Fact Check",
      method: "POST",
      path: "/v1/explanations/{explanation_id}/fact-check",
      description: "Re-run deterministic fact-check validation.",
      defaultPathParams: {
        explanation_id: "expl-abc123"
      }
    },
    {
      id: "explanations-claims",
      title: "Get Claims",
      method: "GET",
      path: "/v1/explanations/{explanation_id}/claims",
      description: "Retrieve claim-level fact-check details.",
      defaultPathParams: {
        explanation_id: "expl-abc123"
      }
    },
    {
      id: "explanations-diff-review",
      title: "Diff Review",
      method: "POST",
      path: "/v1/explanations/diff-review",
      description: "Generate an HTML explanation from a code diff.",
      uxHint: "explanation-html",
      defaultBody: JSON.stringify(
        {
          entity_type: "pull-request",
          entity_id: "pr-42",
          diff_text: "--- a/README.md\n+++ b/README.md\n@@ -1 +1 @@\n-old line\n+new line",
          metadata: {},
          claims: []
        },
        null,
        2
      )
    },
    {
      id: "explanations-plan-review",
      title: "Plan Review",
      method: "POST",
      path: "/v1/explanations/plan-review",
      description: "Generate an HTML explanation from a plan document.",
      uxHint: "explanation-html",
      defaultBody: JSON.stringify(
        {
          entity_type: "workflow",
          entity_id: "hello-flow",
          plan_text: "## Plan\n- Step 1: Review constraints\n- Step 2: Validate outputs",
          metadata: {},
          claims: []
        },
        null,
        2
      )
    },
    {
      id: "explanations-project-recap",
      title: "Project Recap",
      method: "POST",
      path: "/v1/explanations/project-recap",
      description: "Generate an HTML explanation summarising a project milestone.",
      uxHint: "explanation-html",
      defaultBody: JSON.stringify(
        {
          entity_type: "project",
          entity_id: "agent33",
          recap_text: "Phase 26 delivered HTML preview and three new explanation endpoints.",
          highlights: [
            "Diff review endpoint added",
            "Plan review endpoint added",
            "Project recap endpoint added"
          ],
          metadata: {},
          claims: []
        },
        null,
        2
      )
    }
  ]
};
